sgx/urts/src/slog.rs

```diff
@@ -13,12 +20,10 @@ static ENCLAVE_SLOG_LOCATION: RecordLocation = RecordLocation {
     module: "<enclave>",
 };
 
-/// # Safety
-///
-/// This function is marked unsafe due to receiving a raw pointer from the caller.
+// This function is unsafe, if the caller misuses the API it can cause undefined behavior
```

Comment:
1. when you mark a whole function `unsafe`, then there are no `unsafe` blocks in its impl. by making the function not explicitly unsafe, it allows us to use `unsafe` scopes inside what is mostly safe code, which seems good
2. it feels redundant to say `extern "C" unsafe`
3. All our other ocalls follow this pattern

Reply:
this rfc is also relevant -- they want to make it so that `unsafe fn` (function wtih unsafe prototype) still requires `unsafe { ` blocks in its implementation to highlight which parts of its implementations are safe and unsafe

http://rust-lang.github.io/rfcs/2585-unsafe-block-in-unsafe-fn.html

in current versions of the language you can't mark the prototype unsafe without marking the entire implementation unsafe

Reply:
> 1. when you mark a whole function `unsafe`, then there are no `unsafe` blocks in its impl. by making the function not explicitly unsafe, it allows us to use `unsafe` scopes inside what is mostly safe code, which seems good, since less code overall is permitted to do unsafe things, which may ease review

Marking a function `unsafe` affects both the using of unsafe code within the function, as well as whether safe code can call it or not. I agree that the function is mostly safe and by marking the whole function unsafe, it's making it hard to see at a glace which parts of the function are the unsafe bits. However, if we want to separate out the unsafe bits from the safe bits, we should split it up into multiple functions.

The _most important_ reason for marking a function as `unsafe` is so that the compiler emits an error if safe code tries to call it. By marking the function as safe, we'd effectively be opening up a hole in Rust's safety paradigm. Safe code could call this function and could pass it arguments that cause the function to perform unsafe memory access.

As I understand it, marking this function as safe would be a violation of Rust's safety model.

> 2. it feels redundant to say `extern "C" unsafe`

It's not. `extern "C"` refers to symbol mangling and how the symbol is exported. `extern "C"` functions are often also unsafe because they need to access C-compatible arguments, which often means unsafe memory access, but in terms of the Rust language, `extern "C"` does not imply `unsafe`.

Reply:
> The most important reason for marking a function as unsafe is so that the compiler emits an error if safe code tries to call it.

yes but this won't happen when it is extern "C", it is going to be called by C code.

> As I understand it, marking this function as safe would be a violation of Rust's safety model.

Rust's safety model already has a lot of holes, i don't think what we are doing here is a serious risk. If you think it is, then you should open a ticket so that we can review all of our OCALL signatures, i'm just following our established patterns. It's out of scope for this PR to find and change all the rust-side OCALL signatures.

Reply:
> The most important reason for marking a function as unsafe is so that the compiler emits an error if safe code tries to call it. 

I disagree with this, because of the issue mentioned in the RFC. It is easier to review this code when less of it is inside of an unsafe block. It is standard best practice to make unsafe scopes as small as possible, for obvious reasons. What I did in this PR is apply that guideline.

Reply:
basically, in this code, because of the constraints of the language, we have to choose between either
(1) the compiler forcing the rust caller to use "unsafe" appropriately, or
(2) the compiler forcing us to use `unsafe` appropriately inside the function body.

because there is no rust caller, the second is clearly preferable.

I hope they implement the RFC soon.

Reply:
I think @kylefleming is right here: the function should be unsafe, it's an FFI boundary, and it's manipulating raw pointers.

You can have a safe function this chains to which does all the real bits using the slice, but the function here should be unsafe.

Reply:
Yeah, okay.

Do you remember if there are any other ocalls in the system besides in this crate? It was too long ago that I worked on that

Reply:
I guess i'm adjusting my stance, i'm going to try to fix it as suggested in all instances in this crate in this PR

Reply:
> I guess i'm adjusting my stance, i'm going to try to fix it as suggested in all instances in this crate in this PR

I think it would easier to code review if it was a seperate PR.

Reply:
ok, i will do it in a separate PR, thanks